### PR TITLE
Node: Replace usage of Variant with SumType

### LINKED ISSUE #281: Use of std.sumtype.SumType instead of std.variant.Algebraic
Currently `Node` type internally uses [`std.variant.Algebraic`](https://dlang.org/phobos/std_variant.html#.Algebraic) but the official document says it becomes outdated.

Is there a plan to migrate from `Algebraic` to [`std.sumtype.SumType`](https://dlang.org/phobos/std_sumtype.html) or can I send a pull request for it?
IMO it is worth doing because it enables D-YAML to work with attributes such as `pure`.

### PR DESCRIPTION
**Pull request by Geod24**
Fixes #281

### CODE REVIEW

**Comment by Geod24** [comment]
Green

**Comment by Geod24** [comment]
It's a requirement to be able to support DIP1000

File: source/dyaml/node.d [hunk 1]
```diff
@@ -2325,9 +2300,16 @@ struct Node
...
         {
-            return value_.type is typeid(Unqual!T);
+            return value_.match!(
+                (const T _) => true,
+                _    => false,
+            );
         }
 
+        /// Check at compile time if a type is stored natively
+        enum canBeType (T) = is(typeof({ value_.match!((const T _) => true, _ => false); }));
```
Comment (Herringway) [inline source/dyaml/node.d:2310]: What's this for? the comment makes it sound identical in purpose to allowed(T)

File: source/dyaml/node.d [hunk 2]
```diff
@@ -2446,12 +2428,30 @@ struct Node
         // Safe wrapper for getting a value out of the variant.
         inout(T) getValue(T)() @trusted inout
         {
-            return value_.get!T;
```
Comment (Herringway) [inline source/dyaml/node.d:2449]: can we make this method `@safe`? if not, we should probably document why it's `@trusted`

File: source/dyaml/node.d [hunk 3]
```diff
@@ -2446,12 +2428,30 @@ struct Node
...
+            alias RType = typeof(return);
+            return value_.tryMatch!((RType r) => r);
         }
         // Safe wrapper for coercing a value out of the variant.
         inout(T) coerceValue(T)() @trusted inout
         {
-            return (cast(Value)value_).coerce!T;
+            alias RType = typeof(return);
```
Comment (Herringway) [inline source/dyaml/node.d:2437]: Would also like this one to be `@safe` or documented

**Review by Herringway — CHANGES_REQUESTED** [review]
Seems good. Just a few things to address.

File: source/dyaml/node.d (hunk 1 above)
Comment (Geod24) [reply source/dyaml/node.d:2310]: `dstring` will match `allowed` but not `canBeType`.
Before the move to `SumType`, we just relied on `Variant` telling us "no" at runtime, but with `SumType`, we actually hit a `static assert` (a compile-time "no"). To keep the same behavior, I just had to wrap the call to make sure it was only visible to the compiler when `isType` could actually return `true`, not when it wold always return `false`.

File: source/dyaml/node.d (hunk 3 above)
Comment (Geod24) [reply source/dyaml/node.d:2437]: Likewise, I'd recommend to delay opening this can of worm for another PR>

File: source/dyaml/node.d (hunk 2 above)
Comment (Geod24) [reply source/dyaml/node.d:2449]: This one was switched to `@safe`.

File: source/dyaml/node.d (hunk 3 above)
Comment (Geod24) [reply source/dyaml/node.d:2437]: Looks like this can be made `@safe` if we add a `toString` to `Node`. I'll look at it tomorrow.

Comment (Herringway) [reply source/dyaml/node.d:2437]: Sounds good to me.

**Comment by Geod24** [comment]
@Herringway : Updated

**Comment by Geod24** [comment]
Except it doesn't work for LDC, and those error messages are really unhelpful.

**Comment by Herringway** [comment]
it's pretty noisy, but I can see that LDC is deciding that the `ubyte[]` and `SysTime` lambdas in that tryMatch are `@system` for some reason

**Comment by Geod24** [comment]
@Herringway : The bug is related to `inout`. I will reduce it and report it upstream, but for now, can we go ahead with the present version ? It ensures that `match` itself is `@safe` if the delegates are.

**Pull request merged**